lib/src/union_find.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
+#[derive(Clone, Copy)]
+struct Node<T>
+where
+    T: Copy + Eq,
```

Comment:
nit: trait bound can be removed from the type, which I think is generally preferred.